### PR TITLE
dist: add dockerfile image packaging

### PR DESCRIPTION
**Pull request by jakedt**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Cincinnati
+
+Cincinnati is a service to transform update information as modeled in a Docker registry (such as Quay.io) into a graph representation that can be consumed by update clients. It is the spiritual successor to the Omaha protocol used to udate Google Chrome and Google Chrome OS based operating systems.
```
Comment (crawford) [inline README.md]: This isn't very accurate. I'm in favor of leaving off the descriptions for now. I'm in the (slow) process of writing a README which talks about the overall design a little more.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+## Common Usage
```
Comment (crawford) [inline README.md]: Can you call this "Quick Start" instead?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+```sh
+# docker run -d -p 8080:8080 <image> --address 0.0.0.0 --registry https://quay.io --repository <namespace/reponame>
```
Comment (crawford) [inline README.md]: Please drop the `docker run` invocations and just use `cargo run`.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+# docker run -d -p 8081:8081 --entrypoint policy-engine <image> --address 0.0.0.0 --upstream http://localhost:8080/v1/graph
+# curl -v -H 'Accept:application/json' http://localhost:8080/graph
```
Comment (crawford) [inline README.md]: `/v1/graph`

Can you use long flags as well?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+## Components
+
+### Graph Bulder
+
+The Graph Builder is responsible for transforming the data in a Docker registry repository into a directed acyclic graph (DAG) describing update payloads as nodes, and their relationship as edges.
```
Comment (crawford) [inline README.md]: "...(DAG), describing update payloads as nodes and valid update transitions as edges."

File: README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+# curl -v -H 'Accept:application/json' http://localhost:8080/graph
+```
+
+## Components
+
+### Graph Bulder
+
+The Graph Builder is responsible for transforming the data in a Docker registry repository into a directed acyclic graph (DAG) describing update payloads as nodes, and their relationship as edges. 
+
+There is one repository that serves as a command and control repository for each update payload. Each tag in the command and control repository is pulled individually, and each layer is inspected for a `cincinnati.json` file until one is found. The contents of the `cincinnati.json` file describe the update payload for that label, and provide metadata about the update payload itself.
```
Comment (crawford) [inline README.md]: "...for every update payload."

File: README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    ],
+    "metadata": {
+        ...
+    }
+}
+```
+
+### Policy Engine
+
+The Policy Engine used to selectively serve updates to clients. It uses a number of inputs to help shape the graph that is served to each client.
```
Comment (crawford) [inline README.md]: This isn't true yet...

File: README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+## Components
+
+### Graph Bulder
+
+The Graph Builder is responsible for transforming the data in a Docker registry repository into a directed acyclic graph (DAG) describing update payloads as nodes, and their relationship as edges. 
+
+There is one repository that serves as a command and control repository for each update payload. Each tag in the command and control repository is pulled individually, and each layer is inspected for a `cincinnati.json` file until one is found. The contents of the `cincinnati.json` file describe the update payload for that label, and provide metadata about the update payload itself.
+
+### cincinnati.json
```
Comment (crawford) [inline README.md]: I would leave this off. This seems too detailed for a README.

**Review by crawford — CHANGES_REQUESTED** [review]
Can you also fix your commit titles to match the rest of the commits in the history?

File: README.md (hunk 8 above)
Comment (jakedt) [reply README.md]: It has to be documented somewhere... I had to reverse engineer it from the serde definition.

File: README.md (hunk 7 above)
Comment (jakedt) [reply README.md]: I changed it to be technically true but kept the spirit of the description (which I took from the design doc)

File: README.md (hunk 1 above)
Comment (jakedt) [reply README.md]: The repo needs something. I suggest you start with what I have and iterate until it's as nice as you want it to be.

File: README.md (hunk 3 above)
Comment (jakedt) [reply README.md]: My image doesn't package cargo or the rust build system, just the binaries. I would be happy to just reference the binary names.

File: README.md (hunk 1 above)
Comment (jakedt) [reply README.md]: What part of it isn't accurate?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,19 @@
+FROM centos
+
+RUN yum -y groupinstall 'Development Tools'
+RUN yum -y install openssl-devel
+
+ADD https://static.rust-lang.org/dist/rust-1.29.0-x86_64-unknown-linux-gnu.tar.gz rust.tar.gz
```
[the remark this answers is not in the document]
Comment (jakedt) [reply Dockerfile]: AFAICT there are no official Rust packages, and as for rustup, I could have curled it down and run it, but I chose not to use it because I want the builds to be as immutable as possible.

File: README.md (hunk 8 above)
Comment (crawford) [reply README.md]: There are internal documents that detail its structure. I'm in the process of adding these directly to the repo, but it's not as simple as copying and pasting.

**Comment by jakedt** [comment]
Updated to remove `cincinnati.json` example now that the design docs are in.

File: .dockerignore [hunk 10]
```diff
@@ -0,0 +1 @@
+README.md
```
Comment (crawford) [inline .dockerignore]: Why is this needed? I'd like to avoid polluting this repo with a bunch of docker-specific stuff.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
+# Cincinnati
+
+Cincinnati is a service to transform update information as modeled in a Docker registry (such as Quay.io or DockerHub) into a graph representation that can be interpreted by update clients. It is the spiritual successor to the Omaha protocol used to udate Google Chrome and Google Chrome OS based operating systems.
```
Comment (crawford) [inline README.md]: Let's just use the copy from the docs.

"Cincinnati is an update protocol designed to facility automatic updates. It describes a particular method for representing transitions between releases of a project and allowing a client to perform automatic updates between these releases."

File: README.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+## Quick Start
+
+```sh
+# graph-builder --address 0.0.0.0 --registry https://quay.io --repository <namespace/reponame>
```
Comment (crawford) [inline README.md]: `cargo run --package graph-builder -- --address 0.0.0.0 ...`

File: README.md [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+# policy-engine --address 0.0.0.0 --upstream http://localhost:8080/v1/graph
+# curl --verbose --header 'Accept:application/json' http://localhost:8080/v1/graph
+```
+
+## Components
```
Comment (crawford) [inline README.md]: I would drop this section since it is covered in the docs.

File: dist/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,19 @@
+FROM centos
+
+RUN yum -y groupinstall 'Development Tools'
+RUN yum -y install openssl-devel
+
+ADD https://static.rust-lang.org/dist/rust-1.29.0-x86_64-unknown-linux-gnu.tar.gz rust.tar.gz
+# COPY rust-1.29.0-x86_64-unknown-linux-gnu.tar.gz rust.tar.gz
```
Comment (crawford) [inline dist/Dockerfile]: Drop this.

**Review by crawford — CHANGES_REQUESTED** [review]
Just a few final requests.

File: .dockerignore (hunk 10 above)
Comment (crawford) [reply .dockerignore]: Omitting the README from the build context makes development faster? How so? Cache invalidation?

**Comment by jakedt** [comment]
updated

File: README.md [hunk 15]
```diff
@@ -0,0 +1,11 @@
+# Cincinnati
+
+Cincinnati is an update protocol designed to facility automatic updates. It describes a particular method for representing transitions between releases of a project and allowing a client to perform automatic updates between these releases.
+
+## Quick Start
+
+```sh
+# cargo run --package graph-builder -- --address 0.0.0.0 --registry https://quay.io --repository <namespace/reponame>
+# cargo run --package policy-engine -- --address 0.0.0.0 --upstream http://localhost:8080/v1/graph
```
Comment (crawford) [inline README.md]: You can drop the upstream flag here.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+# curl --verbose --header 'Accept:application/json' http://localhost:8080/v1/graph
```
Comment (crawford) [inline README.md]: This should probably be `localhost:8081` if it's the Policy Engine you are trying to hit.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,11 @@
+# Cincinnati
+
+Cincinnati is an update protocol designed to facility automatic updates. It describes a particular method for representing transitions between releases of a project and allowing a client to perform automatic updates between these releases.
+
+## Quick Start
+
+```sh
+# cargo run --package graph-builder -- --address 0.0.0.0 --registry https://quay.io --repository <namespace/reponame>
```
Comment (crawford) [inline README.md]: Please drop the shell prompts. You can use ` ```console` if that's what you are shooting for, but there's no output from these commands. You'll also need to background these invocations if you want this snippet to work.

**Review by crawford — CHANGES_REQUESTED** [review]

**Comment by jakedt** [comment]
updated

**Review by crawford — APPROVED** [review]
